src/libstore/derivations.cc

```diff
@@ -1097,36 +1097,52 @@ Derivation Derivation::fromJSON(
     const Store & store,
     const nlohmann::json & json)
 {
+    using nlohmann::detail::value_t;
+
     Derivation res;
 
-    res.name = json["name"];
+    if (json.type() != value_t::object) {
+        throw Error(
+            "Expected JSON of derivation to be of type '%s', but it is of type '%s'",
+            "object", // Colour this word as well
+            json.type_name());
+    }
```

Comment:
And then split out this function here?

Reply:
I don't think it makes that much sense. The error messages can't be the same.

Reply:
I am saying you have one function for asserting types, and one function for safely looking up keys, but those are orthogonal concerns. Does that make more sense?

Reply:
Right, I see what you're saying. I've moved both of those into separate functions but still kept the combination of both as well, as it aids readability and allows providing more context in the error message. The messages did get longer, see the updated first post, but I think this is alright.